calicoctl/resourcemgr/resourcemgr.go

```diff
@@ -193,28 +201,45 @@ func (rh resourceHelper) Create(client client.Interface, resource runtime.Object
 // We retry 5 times if there is an update conflict during the Update operation.
 func (rh resourceHelper) Update(client client.Interface, resource runtime.Object) (runtime.Object, error) {
 	var err error
-	for i := 0; i < 5; i++ {
+	rv := ""
 
-		// Get the resource to get the resourceVersion.
-		ro, err := rh.get(client, resource.(ResourceObject))
-		if err != nil {
-			return ro, err
-		}
+	// Check to see if the resourceVersion is specified in the resource object.
+	_, ok := resource.(ResourceObject)
+	if ok {
+		rv = resource.(ResourceObject).GetObjectMeta().GetResourceVersion()
+	} else {
+		rv = resource.(ResourceObject).GetObjectMeta().GetResourceVersion()
+	}
 
-		// Set the resource-to-be-updated's resourceVersion to the one we got from the Get call.
-		resource.(ResourceObject).GetObjectMeta().SetResourceVersion(ro.(ResourceObject).GetObjectMeta().GetResourceVersion())
+	// If the resourceVersion is specified then we don't try to Get it or retry.
+	if rv != "" {
+		return rh.update(client, resource.(ResourceObject))
+	} else {
+		// If the resourceVersion is not specified then we do a Get to get
+		// the latest resourceVersion and then do an Update with it.
+		// We retry only if we get an update conflict.
+		for i := 0; i < 5; i++ {
+			// Get the resource to get the resourceVersion.
+			ro, err := rh.get(client, resource.(ResourceObject))
+			if err != nil {
+				return ro, err
+			}
 
-		// Try to update with the resource with the updated resourceVersion.
-		ru, err := rh.update(client, resource.(ResourceObject))
-		if _, ok := err.(cerrors.ErrorResourceUpdateConflict); ok {
-			// Wait for a second and try again if there was a conflict during the resource update.
-			log.Infof("Error updating the resource %s: %s. Retrying.", resource.(ResourceObject).GetObjectMeta().GetName(), err)
-			time.Sleep(1 * time.Second)
-			continue
-		}
+			// Set the resource-to-be-updated's resourceVersion to the one we got from the Get call.
+			resource.(ResourceObject).GetObjectMeta().SetResourceVersion(ro.(ResourceObject).GetObjectMeta().GetResourceVersion())
```

Comment:
Lots of casting to type ResourceObject which could be removed if the method signature just used this type rather than runtime.Object.